{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2aaa0b10-a607-417e-b155-3a5dc0fa4765",
   "metadata": {},
   "outputs": [],
   "source": [
    "#exception class as the base class creates a custom exception needed because \n",
    "#by inheriting from Exceptiion class your custoom exception becomes part of pythons exception heirarchy \n",
    "#exception class provides essential functionality common to all exceptions,such as ability to set an error message etc..\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "77b5a2a9-38e7-4d8f-9c47-6dae43a96858",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'BaseException'>\n",
      "  <class 'Exception'>\n",
      "    <class 'TypeError'>\n",
      "      <class 'decimal.FloatOperation'>\n",
      "      <class 'email.errors.MultipartConversionError'>\n",
      "    <class 'StopAsyncIteration'>\n",
      "    <class 'StopIteration'>\n",
      "    <class 'ImportError'>\n",
      "      <class 'ModuleNotFoundError'>\n",
      "      <class 'zipimport.ZipImportError'>\n",
      "    <class 'OSError'>\n",
      "      <class 'ConnectionError'>\n",
      "        <class 'BrokenPipeError'>\n",
      "        <class 'ConnectionAbortedError'>\n",
      "        <class 'ConnectionRefusedError'>\n",
      "        <class 'ConnectionResetError'>\n",
      "          <class 'http.client.RemoteDisconnected'>\n",
      "      <class 'BlockingIOError'>\n",
      "      <class 'ChildProcessError'>\n",
      "      <class 'FileExistsError'>\n",
      "      <class 'FileNotFoundError'>\n",
      "      <class 'IsADirectoryError'>\n",
      "      <class 'NotADirectoryError'>\n",
      "      <class 'InterruptedError'>\n",
      "        <class 'zmq.error.InterruptedSystemCall'>\n",
      "      <class 'PermissionError'>\n",
      "      <class 'ProcessLookupError'>\n",
      "      <class 'TimeoutError'>\n",
      "      <class 'io.UnsupportedOperation'>\n",
      "      <class 'signal.itimer_error'>\n",
      "      <class 'socket.herror'>\n",
      "      <class 'socket.gaierror'>\n",
      "      <class 'ssl.SSLError'>\n",
      "        <class 'ssl.SSLCertVerificationError'>\n",
      "        <class 'ssl.SSLZeroReturnError'>\n",
      "        <class 'ssl.SSLWantWriteError'>\n",
      "        <class 'ssl.SSLWantReadError'>\n",
      "        <class 'ssl.SSLSyscallError'>\n",
      "        <class 'ssl.SSLEOFError'>\n",
      "      <class 'shutil.Error'>\n",
      "        <class 'shutil.SameFileError'>\n",
      "      <class 'shutil.SpecialFileError'>\n",
      "      <class 'shutil.ExecError'>\n",
      "      <class 'shutil.ReadError'>\n",
      "      <class 'urllib.error.URLError'>\n",
      "        <class 'urllib.error.HTTPError'>\n",
      "        <class 'urllib.error.ContentTooShortError'>\n",
      "      <class 'gzip.BadGzipFile'>\n",
      "    <class 'EOFError'>\n",
      "      <class 'asyncio.exceptions.IncompleteReadError'>\n",
      "    <class 'RuntimeError'>\n",
      "      <class 'RecursionError'>\n",
      "      <class 'NotImplementedError'>\n",
      "        <class 'zmq.error.ZMQVersionError'>\n",
      "        <class 'IPython.core.error.StdinNotImplementedError'>\n",
      "      <class '_frozen_importlib._DeadlockError'>\n",
      "      <class 'threading.BrokenBarrierError'>\n",
      "      <class 'concurrent.futures._base.BrokenExecutor'>\n",
      "        <class 'concurrent.futures.thread.BrokenThreadPool'>\n",
      "      <class 'asyncio.exceptions.SendfileNotAvailableError'>\n",
      "      <class 'pkg_resources.ExtractionError'>\n",
      "      <class '_pydevd_bundle.pydevd_vars.VariableError'>\n",
      "    <class 'NameError'>\n",
      "      <class 'UnboundLocalError'>\n",
      "    <class 'AttributeError'>\n",
      "      <class 'dataclasses.FrozenInstanceError'>\n",
      "    <class 'SyntaxError'>\n",
      "      <class 'IndentationError'>\n",
      "        <class 'TabError'>\n",
      "    <class 'LookupError'>\n",
      "      <class 'IndexError'>\n",
      "      <class 'KeyError'>\n",
      "        <class 'jupyter_client.kernelspec.NoSuchKernel'>\n",
      "        <class 'IPython.terminal.pt_inputhooks.UnknownBackend'>\n",
      "      <class 'encodings.CodecRegistryError'>\n",
      "    <class 'ValueError'>\n",
      "      <class 'UnicodeError'>\n",
      "        <class 'UnicodeEncodeError'>\n",
      "        <class 'UnicodeDecodeError'>\n",
      "        <class 'UnicodeTranslateError'>\n",
      "      <class 'io.UnsupportedOperation'>\n",
      "      <class 'json.decoder.JSONDecodeError'>\n",
      "      <class 'ssl.SSLCertVerificationError'>\n",
      "      <class 'binascii.Error'>\n",
      "      <class '_hashlib.UnsupportedDigestmodError'>\n",
      "      <class 'calendar.IllegalMonthError'>\n",
      "      <class 'calendar.IllegalWeekdayError'>\n",
      "      <class 'dateutil.parser._parser.ParserError'>\n",
      "      <class 'pygments.util.ClassNotFound'>\n",
      "      <class 'IPython.lib.clipboard.ClipboardEmpty'>\n",
      "      <class 'email.errors.MessageDefect'>\n",
      "        <class 'email.errors.NoBoundaryInMultipartDefect'>\n",
      "        <class 'email.errors.StartBoundaryNotFoundDefect'>\n",
      "        <class 'email.errors.CloseBoundaryNotFoundDefect'>\n",
      "        <class 'email.errors.FirstHeaderLineIsContinuationDefect'>\n",
      "        <class 'email.errors.MisplacedEnvelopeHeaderDefect'>\n",
      "        <class 'email.errors.MissingHeaderBodySeparatorDefect'>\n",
      "        <class 'email.errors.MultipartInvariantViolationDefect'>\n",
      "        <class 'email.errors.InvalidMultipartContentTransferEncodingDefect'>\n",
      "        <class 'email.errors.UndecodableBytesDefect'>\n",
      "        <class 'email.errors.InvalidBase64PaddingDefect'>\n",
      "        <class 'email.errors.InvalidBase64CharactersDefect'>\n",
      "        <class 'email.errors.InvalidBase64LengthDefect'>\n",
      "        <class 'email.errors.HeaderDefect'>\n",
      "          <class 'email.errors.InvalidHeaderDefect'>\n",
      "          <class 'email.errors.HeaderMissingRequiredValue'>\n",
      "          <class 'email.errors.NonPrintableDefect'>\n",
      "          <class 'email.errors.ObsoleteHeaderDefect'>\n",
      "          <class 'email.errors.NonASCIILocalPartDefect'>\n",
      "          <class 'email.errors.InvalidDateDefect'>\n",
      "      <class 'IPython.core.magics.code.MacroToEdit'>\n",
      "      <class 'plistlib.InvalidFileException'>\n",
      "      <class 'pkg_resources._vendor.more_itertools.more.UnequalIterablesError'>\n",
      "      <class 'pkg_resources.extern.packaging.version.InvalidVersion'>\n",
      "      <class 'pkg_resources._vendor.packaging._manylinux._ELFFileHeader._InvalidELFFileHeader'>\n",
      "      <class 'pkg_resources.extern.packaging.utils.InvalidWheelFilename'>\n",
      "      <class 'pkg_resources.extern.packaging.utils.InvalidSdistFilename'>\n",
      "      <class 'pkg_resources.extern.packaging.specifiers.InvalidSpecifier'>\n",
      "      <class 'pkg_resources.extern.packaging.markers.InvalidMarker'>\n",
      "      <class 'pkg_resources.extern.packaging.markers.UndefinedComparison'>\n",
      "      <class 'pkg_resources.extern.packaging.markers.UndefinedEnvironmentName'>\n",
      "      <class 'pkg_resources.extern.packaging.requirements.InvalidRequirement'>\n",
      "        <class 'pkg_resources.RequirementParseError'>\n",
      "      <class 'packaging.version.InvalidVersion'>\n",
      "    <class 'AssertionError'>\n",
      "    <class 'ArithmeticError'>\n",
      "      <class 'FloatingPointError'>\n",
      "      <class 'OverflowError'>\n",
      "      <class 'ZeroDivisionError'>\n",
      "        <class 'decimal.DivisionByZero'>\n",
      "        <class 'decimal.DivisionUndefined'>\n",
      "      <class 'decimal.DecimalException'>\n",
      "        <class 'decimal.Clamped'>\n",
      "        <class 'decimal.Rounded'>\n",
      "          <class 'decimal.Underflow'>\n",
      "          <class 'decimal.Overflow'>\n",
      "        <class 'decimal.Inexact'>\n",
      "          <class 'decimal.Underflow'>\n",
      "          <class 'decimal.Overflow'>\n",
      "        <class 'decimal.Subnormal'>\n",
      "          <class 'decimal.Underflow'>\n",
      "        <class 'decimal.DivisionByZero'>\n",
      "        <class 'decimal.FloatOperation'>\n",
      "        <class 'decimal.InvalidOperation'>\n",
      "          <class 'decimal.ConversionSyntax'>\n",
      "          <class 'decimal.DivisionImpossible'>\n",
      "          <class 'decimal.DivisionUndefined'>\n",
      "          <class 'decimal.InvalidContext'>\n",
      "    <class 'SystemError'>\n",
      "      <class 'encodings.CodecRegistryError'>\n",
      "    <class 'ReferenceError'>\n",
      "    <class 'MemoryError'>\n",
      "    <class 'BufferError'>\n",
      "    <class 'Warning'>\n",
      "      <class 'UserWarning'>\n",
      "        <class 'getpass.GetPassWarning'>\n",
      "        <class 'IPython.core.formatters.FormatterWarning'>\n",
      "      <class 'EncodingWarning'>\n",
      "      <class 'DeprecationWarning'>\n",
      "        <class 'IPython.core.interactiveshell.ProvisionalWarning'>\n",
      "      <class 'PendingDeprecationWarning'>\n",
      "      <class 'SyntaxWarning'>\n",
      "      <class 'RuntimeWarning'>\n",
      "        <class 'zmq.asyncio.ProactorSelectorThreadWarning'>\n",
      "        <class 'dateutil.parser._parser.UnknownTimezoneWarning'>\n",
      "        <class 'pkg_resources.PEP440Warning'>\n",
      "      <class 'FutureWarning'>\n",
      "        <class 'IPython.core.completer.ProvisionalCompleterWarning'>\n",
      "      <class 'ImportWarning'>\n",
      "      <class 'UnicodeWarning'>\n",
      "      <class 'BytesWarning'>\n",
      "      <class 'ResourceWarning'>\n",
      "      <class 'dateutil.tz.DeprecatedTzFormatWarning'>\n",
      "      <class 'pkg_resources.PkgResourcesDeprecationWarning'>\n",
      "    <class 'warnings._OptionError'>\n",
      "    <class 'runpy._Error'>\n",
      "    <class 're.error'>\n",
      "    <class 'sre_parse.Verbose'>\n",
      "    <class 'locale.Error'>\n",
      "    <class 'subprocess.SubprocessError'>\n",
      "      <class 'subprocess.CalledProcessError'>\n",
      "      <class 'subprocess.TimeoutExpired'>\n",
      "    <class 'tokenize.TokenError'>\n",
      "    <class 'tokenize.StopTokenizing'>\n",
      "    <class 'inspect.ClassFoundException'>\n",
      "    <class 'inspect.EndOfBlock'>\n",
      "    <class 'traitlets.traitlets.TraitError'>\n",
      "    <class 'copy.Error'>\n",
      "    <class 'concurrent.futures._base.Error'>\n",
      "      <class 'concurrent.futures._base.CancelledError'>\n",
      "      <class 'concurrent.futures._base.TimeoutError'>\n",
      "      <class 'concurrent.futures._base.InvalidStateError'>\n",
      "    <class 'socket._GiveupOnSendfile'>\n",
      "    <class 'struct.error'>\n",
      "    <class 'binascii.Incomplete'>\n",
      "    <class 'asyncio.exceptions.TimeoutError'>\n",
      "    <class 'asyncio.exceptions.InvalidStateError'>\n",
      "    <class 'asyncio.exceptions.LimitOverrunError'>\n",
      "    <class 'asyncio.queues.QueueEmpty'>\n",
      "    <class 'asyncio.queues.QueueFull'>\n",
      "    <class '_queue.Empty'>\n",
      "    <class 'queue.Full'>\n",
      "    <class 'ctypes.ArgumentError'>\n",
      "    <class 'zmq.error.ZMQBaseError'>\n",
      "      <class 'zmq.error.ZMQError'>\n",
      "        <class 'zmq.error.ContextTerminated'>\n",
      "        <class 'zmq.error.Again'>\n",
      "        <class 'zmq.error.InterruptedSystemCall'>\n",
      "      <class 'zmq.error.ZMQBindError'>\n",
      "      <class 'zmq.error.NotDone'>\n",
      "    <class '_pickle.PickleError'>\n",
      "      <class '_pickle.PicklingError'>\n",
      "      <class '_pickle.UnpicklingError'>\n",
      "    <class 'pickle._Stop'>\n",
      "    <class 'argparse.ArgumentError'>\n",
      "    <class 'argparse.ArgumentTypeError'>\n",
      "    <class 'traitlets.config.loader.ConfigError'>\n",
      "      <class 'traitlets.config.loader.ConfigLoaderError'>\n",
      "        <class 'traitlets.config.loader.ArgumentError'>\n",
      "      <class 'traitlets.config.loader.ConfigFileNotFound'>\n",
      "    <class 'traitlets.config.configurable.ConfigurableError'>\n",
      "      <class 'traitlets.config.configurable.MultipleInstanceError'>\n",
      "    <class 'traitlets.config.application.ApplicationError'>\n",
      "    <class 'zlib.error'>\n",
      "    <class 'tornado.util.TimeoutError'>\n",
      "    <class '_curses.error'>\n",
      "    <class 'tornado.concurrent.ReturnValueIgnoredError'>\n",
      "    <class 'tornado.gen.KeyReuseError'>\n",
      "    <class 'tornado.gen.UnknownKeyError'>\n",
      "    <class 'tornado.gen.LeakedCallbackError'>\n",
      "    <class 'tornado.gen.BadYieldError'>\n",
      "    <class 'tornado.gen.ReturnValueIgnoredError'>\n",
      "    <class 'tornado.gen.Return'>\n",
      "    <class 'jupyter_client.channels.InvalidPortNumber'>\n",
      "    <class 'termios.error'>\n",
      "    <class '_lzma.LZMAError'>\n",
      "    <class 'shutil.RegistryError'>\n",
      "    <class 'shutil._GiveupOnFastCopy'>\n",
      "    <class 'configparser.Error'>\n",
      "      <class 'configparser.NoSectionError'>\n",
      "      <class 'configparser.DuplicateSectionError'>\n",
      "      <class 'configparser.DuplicateOptionError'>\n",
      "      <class 'configparser.NoOptionError'>\n",
      "      <class 'configparser.InterpolationError'>\n",
      "        <class 'configparser.InterpolationMissingOptionError'>\n",
      "        <class 'configparser.InterpolationSyntaxError'>\n",
      "        <class 'configparser.InterpolationDepthError'>\n",
      "      <class 'configparser.ParsingError'>\n",
      "        <class 'configparser.MissingSectionHeaderError'>\n",
      "    <class 'jupyter_client.localinterfaces.NoIPAddresses'>\n",
      "    <class 'zipfile.BadZipFile'>\n",
      "    <class 'zipfile.LargeZipFile'>\n",
      "    <class 'entrypoints.BadEntryPoint'>\n",
      "    <class 'entrypoints.NoSuchEntryPoint'>\n",
      "    <class 'jupyter_client.multikernelmanager.DuplicateKernelError'>\n",
      "    <class 'pydoc.ErrorDuringImport'>\n",
      "    <class 'executing.executing.NotOneValueFound'>\n",
      "    <class 'pure_eval.utils.CannotEval'>\n",
      "    <class 'pygments.util.OptionError'>\n",
      "    <class 'bdb.BdbQuit'>\n",
      "    <class 'pdb.Restart'>\n",
      "    <class 'pexpect.exceptions.ExceptionPexpect'>\n",
      "      <class 'pexpect.exceptions.EOF'>\n",
      "      <class 'pexpect.exceptions.TIMEOUT'>\n",
      "    <class 'ptyprocess.util.PtyProcessError'>\n",
      "    <class 'IPython.utils.process.FindCmdError'>\n",
      "    <class 'IPython.utils.path.HomeDirError'>\n",
      "    <class 'IPython.core.profiledir.ProfileDirError'>\n",
      "    <class 'IPython.core.error.IPythonCoreError'>\n",
      "      <class 'IPython.core.error.TryNext'>\n",
      "      <class 'IPython.core.error.UsageError'>\n",
      "      <class 'IPython.core.error.StdinNotImplementedError'>\n",
      "    <class 'IPython.core.error.InputRejected'>\n",
      "    <class 'getopt.GetoptError'>\n",
      "    <class 'pygments.filters.ErrorToken'>\n",
      "    <class 'IPython.core.prefilter.PrefilterError'>\n",
      "    <class 'IPython.core.alias.AliasError'>\n",
      "      <class 'IPython.core.alias.InvalidAliasError'>\n",
      "    <class 'sqlite3.Error'>\n",
      "      <class 'sqlite3.InterfaceError'>\n",
      "      <class 'sqlite3.DatabaseError'>\n",
      "        <class 'sqlite3.InternalError'>\n",
      "        <class 'sqlite3.OperationalError'>\n",
      "        <class 'sqlite3.ProgrammingError'>\n",
      "        <class 'sqlite3.IntegrityError'>\n",
      "        <class 'sqlite3.DataError'>\n",
      "        <class 'sqlite3.NotSupportedError'>\n",
      "    <class 'sqlite3.Warning'>\n",
      "    <class 'IPython.core.interactiveshell.SpaceInInput'>\n",
      "    <class 'xml.dom.DOMException'>\n",
      "      <class 'xml.dom.IndexSizeErr'>\n",
      "      <class 'xml.dom.DomstringSizeErr'>\n",
      "      <class 'xml.dom.HierarchyRequestErr'>\n",
      "      <class 'xml.dom.WrongDocumentErr'>\n",
      "      <class 'xml.dom.InvalidCharacterErr'>\n",
      "      <class 'xml.dom.NoDataAllowedErr'>\n",
      "      <class 'xml.dom.NoModificationAllowedErr'>\n",
      "      <class 'xml.dom.NotFoundErr'>\n",
      "      <class 'xml.dom.NotSupportedErr'>\n",
      "      <class 'xml.dom.InuseAttributeErr'>\n",
      "      <class 'xml.dom.InvalidStateErr'>\n",
      "      <class 'xml.dom.SyntaxErr'>\n",
      "      <class 'xml.dom.InvalidModificationErr'>\n",
      "      <class 'xml.dom.NamespaceErr'>\n",
      "      <class 'xml.dom.InvalidAccessErr'>\n",
      "      <class 'xml.dom.ValidationErr'>\n",
      "    <class 'prompt_toolkit.validation.ValidationError'>\n",
      "    <class 'prompt_toolkit.buffer.EditReadOnlyBuffer'>\n",
      "    <class 'prompt_toolkit.buffer._Retry'>\n",
      "    <class 'prompt_toolkit.layout.layout.InvalidLayoutError'>\n",
      "    <class 'prompt_toolkit.renderer.HeightIsUnknownError'>\n",
      "    <class 'parso.parser.ParserSyntaxError'>\n",
      "    <class 'parso.parser.InternalParseError'>\n",
      "    <class 'parso.python.diff._PositionUpdatingFinished'>\n",
      "    <class 'jedi.inference.helpers.SimpleGetItemNotFound'>\n",
      "    <class 'jedi.inference.utils.UncaughtAttributeError'>\n",
      "    <class 'jedi.inference.base_value.HasNoContext'>\n",
      "    <class 'jedi.inference.arguments.ParamIssue'>\n",
      "    <class 'jedi.api.exceptions._JediError'>\n",
      "      <class 'jedi.api.exceptions.InternalError'>\n",
      "      <class 'jedi.api.exceptions.WrongVersion'>\n",
      "      <class 'jedi.api.exceptions.RefactoringError'>\n",
      "    <class 'jedi.api.helpers.OnErrorLeaf'>\n",
      "    <class 'jedi.api.environment.InvalidPythonEnvironment'>\n",
      "    <class 'email.errors.MessageError'>\n",
      "      <class 'email.errors.MessageParseError'>\n",
      "        <class 'email.errors.HeaderParseError'>\n",
      "        <class 'email.errors.BoundaryError'>\n",
      "      <class 'email.errors.MultipartConversionError'>\n",
      "      <class 'email.errors.CharsetError'>\n",
      "    <class 'uu.Error'>\n",
      "    <class 'http.client.HTTPException'>\n",
      "      <class 'http.client.NotConnected'>\n",
      "      <class 'http.client.InvalidURL'>\n",
      "      <class 'http.client.UnknownProtocol'>\n",
      "      <class 'http.client.UnknownTransferEncoding'>\n",
      "      <class 'http.client.UnimplementedFileMode'>\n",
      "      <class 'http.client.IncompleteRead'>\n",
      "      <class 'http.client.ImproperConnectionState'>\n",
      "        <class 'http.client.CannotSendRequest'>\n",
      "        <class 'http.client.CannotSendHeader'>\n",
      "        <class 'http.client.ResponseNotReady'>\n",
      "      <class 'http.client.BadStatusLine'>\n",
      "        <class 'http.client.RemoteDisconnected'>\n",
      "      <class 'http.client.LineTooLong'>\n",
      "    <class 'IPython.core.magics.code.InteractivelyDefined'>\n",
      "    <class 'IPython.terminal.embed.KillEmbedded'>\n",
      "    <class 'psutil.Error'>\n",
      "      <class 'psutil.NoSuchProcess'>\n",
      "        <class 'psutil.ZombieProcess'>\n",
      "      <class 'psutil.AccessDenied'>\n",
      "      <class 'psutil.TimeoutExpired'>\n",
      "    <class 'psutil._pslinux._Ipv6UnsupportedError'>\n",
      "    <class 'tornado.queues.QueueEmpty'>\n",
      "    <class 'tornado.queues.QueueFull'>\n",
      "    <class '_pydev_bundle._pydev_saved_modules.DebuggerInitializationError'>\n",
      "    <class 'xml.parsers.expat.ExpatError'>\n",
      "    <class 'xmlrpc.client.Error'>\n",
      "      <class 'xmlrpc.client.ProtocolError'>\n",
      "      <class 'xmlrpc.client.ResponseError'>\n",
      "      <class 'xmlrpc.client.Fault'>\n",
      "    <class 'pkg_resources._vendor.pyparsing.exceptions.ParseBaseException'>\n",
      "      <class 'pkg_resources._vendor.pyparsing.exceptions.ParseException'>\n",
      "      <class 'pkg_resources._vendor.pyparsing.exceptions.ParseFatalException'>\n",
      "        <class 'pkg_resources._vendor.pyparsing.exceptions.ParseSyntaxException'>\n",
      "    <class 'pkg_resources._vendor.pyparsing.exceptions.RecursiveGrammarException'>\n",
      "    <class 'pkg_resources.ResolutionError'>\n",
      "      <class 'pkg_resources.VersionConflict'>\n",
      "        <class 'pkg_resources.ContextualVersionConflict'>\n",
      "      <class 'pkg_resources.DistributionNotFound'>\n",
      "      <class 'pkg_resources.UnknownExtra'>\n",
      "    <class '_pydevd_bundle.pydevd_runpy._Error'>\n",
      "    <class '_pydevd_bundle.pydevd_resolver.UnableToResolveVariableException'>\n",
      "    <class '_pydev_bundle.pydev_monkey.InvalidTypeInArgsException'>\n",
      "  <class 'GeneratorExit'>\n",
      "  <class 'SystemExit'>\n",
      "  <class 'KeyboardInterrupt'>\n",
      "  <class 'asyncio.exceptions.CancelledError'>\n",
      "  <class 'pkg_resources._vendor.more_itertools.more.AbortThread'>\n"
     ]
    }
   ],
   "source": [
    "def print_exception_hierarchy(exception_class,indent=0):\n",
    "    print(' '*indent +str(exception_class))\n",
    "    for i in exception_class.__subclasses__():\n",
    "        print_exception_hierarchy(i,indent+2)\n",
    "if __name__ == \"__main__\":\n",
    "    print_exception_hierarchy(BaseException)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e305bb66-e1d0-44bc-8515-5e70b8075ee8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "inf\n"
     ]
    }
   ],
   "source": [
    "#FlotingPointError , OverFlowError,ZeroDivisionError are the errors defined in arthmatic error class \n",
    "#floating point error occuurs when a when a program tries to perform an action with floating numbers but encounters an issue that prevents the actionn to compete\n",
    "big = 1.0e1000\n",
    "b = big*2\n",
    "print(b)\n",
    "#here it gives inf as a result but it is not the exact ans this type of exceptions are floating poinn errors\n",
    "10/0\n",
    "#here we here we get a zero division eror  as we are trying to divide a number with 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f765492c-d690-494d-b9c9-ea15d3b998a9",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "3",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[6], line 4\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m#lookup error class is a base class for index error and key error\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;66;03m#keyerror occurs when we try to access a element in a dictionary that is not available in that dictionnary\u001b[39;00m\n\u001b[1;32m      3\u001b[0m d \u001b[38;5;241m=\u001b[39m {\u001b[38;5;241m1\u001b[39m:\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mone\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;241m2\u001b[39m:\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtwo\u001b[39m\u001b[38;5;124m'\u001b[39m}\n\u001b[0;32m----> 4\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[43md\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m3\u001b[39;49m\u001b[43m]\u001b[49m)\u001b[38;5;66;03m#keyerror\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m#index error occurs when you try to access an element in a list that is having a index grater than no of elements in the list\u001b[39;00m\n\u001b[1;32m      6\u001b[0m l \u001b[38;5;241m=\u001b[39m [\u001b[38;5;241m1\u001b[39m,\u001b[38;5;241m2\u001b[39m,\u001b[38;5;241m3\u001b[39m,\u001b[38;5;241m4\u001b[39m]\n",
      "\u001b[0;31mKeyError\u001b[0m: 3"
     ]
    }
   ],
   "source": [
    "#lookup error class is a base class for index error and key error\n",
    "#keyerror occurs when we try to access a element in a dictionary that is not available in that dictionnary\n",
    "d = {1:'one',2:'two'}\n",
    "print(d[3])#keyerror\n",
    "#index error occurs when you try to access an element in a list that is having a index grater than no of elements in the list\n",
    "l = [1,2,3,4]\n",
    "print(l[7])#index error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2d6d3288-f8bc-4cae-8f81-050ff5164264",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'tilak'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 4\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m#import is used to import modules into your python file\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;66;03m#import error occurs when you are trying to import a modele that does not exist\u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;03m#this is results in import error ModuleNotFoundError\u001b[39;00m\n\u001b[0;32m----> 4\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mtilak\u001b[39;00m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'tilak'"
     ]
    }
   ],
   "source": [
    "#import is used to import modules into your python file\n",
    "#import error occurs when you are trying to import a modele that does not exist\n",
    "#this is results in import error ModuleNotFoundError\n",
    "import tilak"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "daa7e4ff-478b-4764-89fd-6d2ae9586dc7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "division by zero\n",
      "got a file error [Errno 2] No such file or directory: 'tilak.txt'\n"
     ]
    }
   ],
   "source": [
    "#best use of exceptiion handling is use of best hierarchy\n",
    "try: \n",
    "    10/0\n",
    "except Exception as e:\n",
    "    print(e)\n",
    "#here we are managing the exception directly using the base Exception class this is not a good practise\n",
    "try:\n",
    "    f = open('tilak.txt','r')\n",
    "    f.write(\"hi\")\n",
    "except AttributeError as e:\n",
    "    print('got attribute error')\n",
    "except NameError as e:\n",
    "    print('got a name error',e)\n",
    "except FileNotFoundError as e:\n",
    "    print('got a file error',e)\n",
    "except Exception as e:\n",
    "    print('error',e)\n",
    "#here we are manaigng a exeption a hierarcy so that the error can be managed in a good order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9710d19-42c8-449b-aba2-b416bada3fc1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
